### ISSUE TITLE
Feature Request: Forms Validation

### ISSUE THREAD

**Issue by DrYSG** [ctrlplusb/easy-peasy#203]
Some folks (see:  https://stackoverflow.com/questions/54743324/use-redux-form-with-easy-peasy)
have been thinking or  trying out how to use the redux store that easy-peasy creates as place to store forms info, and then do validation against it.

As you can see from SO article, that is not working.

I also looked at Formik (but that keeps all the forms info local, not in the STORE, which is not good for us, since we have lots of  server data that provides context-filters and validation cross-checks against the user input.

Has anyone done any thinking about ways to go about this? 

Yea, this is not an issue, but SO does not have a big easy-peasy community, (though I would love to grow it, since I think this is a fantastic tool).

**Comment by DrYSG** [comment]
I think that this could be done with the current **selector**  function. One could then use stateless functional components that put their data and get the validation state from the Redux store. In addition any remote data that contextualizes the validity of the form (e.g. you need a particular security level to fill out fields, you don't have training to ask for that operation, etc.) would also be in the section of the **store**

The idea would be to use **selectors**, so that whenever the user or remote processes update the store, validation values would be returned by the selectors.

What would be nice if the **listen** option was also available to selectors as it is with **thunks**, so that we could have control of whether the checks are run, or whether they have to continuously give valid/invalid values.

**Comment by allpro** [comment]
It would be difficult to add form handling logic versatile enough to handle all usecases. For example, you do not always want to validate on every keystroke, nor do you want ALL validation done inside the store. That would be cumbersome.

My FormManager helper would be easy to integrate with validators in the store. It's vastly for powerful and extensible than Formik. Custom validators can be easily attached to any field, in combination with builtin validators. FormManager handles all the logic and flow, such as when to validate, and passing through errors returned from state to the presentation component.

FormManager has a React Hook that would live in the same container as the state hooks, so it's simple for these two 'data-handlers' to interoperate. It's probably unnecessary to push EVERY change into the store. Just integrate the store aspects you need and let FormManager do the rest. This would dramatically simplify your code and props.

The documentation is a work in progress, but I could show you how to do what you want if you provide an example.

https://www.npmjs.com/package/@allpro/form-manager
https://allpro.github.io/form-manager

**Comment by ctrlplusb** [comment]
Hey @DrYSG 

Thanks for raising this. 

To be honest I am strongly against storing form state in a global store. I find it is too transient to be considered global state. In my opinion it pollutes your global state and can make it quite hard to parse your redux dev tools, inhibiting some debugging capabilities.

I appreciate not every case is the same though, so I respect it if you have some sort of requirement that forces you to put form state in the global store. That being said I would be interested if you could create some sort of illustrative minimal codesandbox example. Doing this may open up the door for myself and the community to be able to review your architectural design and perhaps make alternative suggestions.

I'll leave this issue open for now in the hope someone else may have some more immediate insights for you.

**Comment by ctrlplusb** [comment]
Closing this now due to inactivity. 👍

**Comment by DrYSG** [comment]
@ctrlplusb Seth, I see where you are coming from. I do a lot of Big R, little D (R&d) so we do some really strange things here.  We like to stretch things to break them. So in our case, having all the logic for state management across forms,, devices, cloud stream services, etc. makes life a lot cleaner.
